---
title: "SCHEMA analysis"
output: html_document
date: "2023-09-28"
---


```{r, echo=F, warning=F, message=F}
rm(list=ls())
set.seed(123)
library(knitr)
library(RSQLite)
library(dplyr)
library(knitr)
knitr::opts_chunk$set(cache = TRUE, warning = FALSE, 
                      message = FALSE, cache.lazy = FALSE) # avoid error of long vectors https://stackoverflow.com/questions/39417003/long-vectors-not-supported-yet-error-in-rmd-but-not-in-r-script 
library(kableExtra)
library(RColorBrewer)
library(gplots)
library(tidyverse)
library(gridExtra)
library(ggpubr)
library(DT)
#library("devtools")
#install.packages("Rtools")
#install_github('xinhe-lab/mirage')
library(mirage)

multiplesheets <- function(fname) {
   
  # getting info about all excel sheets
  sheets <- readxl::excel_sheets(fname)
  tibble <- lapply(sheets, function(x) readxl::read_excel(fname, sheet = x))
  data_frame <- lapply(tibble, as.data.frame)
    
  # assigning names to data frames
  names(data_frame) <- sheets
    
  # print data frame
  print(data_frame)
}


```

[SCHEMA](https://schema.broadinstitute.org/)

```{r, echo=F, message=F, warning=F,cache=T}
SCHEMA_data=as_tibble(read.table("C:\\han\\Dataset\\SCHEMA\\SCHEMA_variant_results_subset_20231016.tsv", header=T))
```


```{r, echo=F, message=F, warning=F}
# Conversion of Gene Name to Ensembl ID https://www.biostars.org/p/430015/

# install package biomaRT
#if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

#BiocManager::install("biomaRt")

library(biomaRt)
mart <- useMart('ENSEMBL_MART_ENSEMBL')
mart <- useDataset('hsapiens_gene_ensembl', mart)

annotLookup <- getBM(
  mart = mart,
  attributes = c(
    'hgnc_symbol',
    'ensembl_gene_id',
    'gene_biotype'),
  uniqueRows = TRUE)
colnames(annotLookup)[2]="gene_id"

SCHEMA_data_with_hgnc_symbol=right_join(annotLookup, SCHEMA_data, by="gene_id") 
```



## Replicate Extended Data Fig 1

```{r, echo=F, message=F, warning=F, results=F}
Table13=multiplesheets("../../../../Dataset/20231017_PMID_27535533/Lek 2016 suppl table/nature19057-s2/nature19057-SI Table 13.xlsx")

suppl_table11=read.table(gzfile("../../../../Dataset/20231018_PMID_32461654/41586_2020_2308_MOESM4_ESM/supplement/supplementary_dataset_11_full_constraint_metrics.tsv.gz"),fill = TRUE , header = TRUE ) # read into gz file 
# length(unique(suppl_table11 %>% filter(pLI>0.9) %>% select(gene) %>%pull)) 3476 genes with pLI>0.9 
```

```{r, echo=F, message=F, warning=F}
LoF_intolerant=Table13$`LoF Intolerant`
# length(unique(LoF_intolerant %>% filter(pLI>0.9) %>% (gene) %>% pull)) # 3230 genes with pLI>0.9, consistent with the number reported in paper Lek2017_Nature 

LoF_intolerant_gene=unique(suppl_table11 %>% filter(pLI>0.9) %>% dplyr::select(gene) %>%pull) 

cohorts=c("meta","UK_Ireland_EUR", "Swedish_EUR",  "Danish_EUR", "UK10K_NFBC_FINRISK_FIN", "UK10K_INTERVAL_EUR", "GPC_Latino_AMR", "GPC_McLean_EUR", "GPC_genomes_AFR", "ASJ",  "EAS", "EST", "Bulgarian_EUR", "FIN", "GPC_genomes_EUR", "Swedish_FIN", "GPC_genomes_AMR")
PTV_definition=c("frameshift_variant", "stop_gained", "splice_acceptor_variant","splice_donor_variant" )
odds_ratio_summary_ptv=matrix(nrow=length(cohorts), ncol=5)
odds_ratio_summary_syn=matrix(nrow=length(cohorts), ncol=5)
num_case=22444; num_ctrl=39837

MAC_threshold=5 # this cutoff is used in SCHEMA paper in Section Analysis approach
#for (i in 1:length(cohorts))
{
  i=2
  cohort_ptv=SCHEMA_data_with_hgnc_symbol %>% filter(hgnc_symbol %in%LoF_intolerant_gene==T &  consequence %in% PTV_definition==T & group==cohorts[i] & ac_case<=MAC_threshold & ac_ctrl<=MAC_threshold)
  cohort_syn=SCHEMA_data_with_hgnc_symbol %>% filter(hgnc_symbol %in%LoF_intolerant_gene==T &  consequence== "synonymous_variant" & group==cohorts[i] & ac_case<=MAC_threshold & ac_ctrl<=MAC_threshold)
  
  ptv_burden_test=fisher.test(matrix(c(sum(cohort_ptv$ac_case),num_case,  sum(cohort_ptv$ac_ctrl), num_ctrl), nrow=2))
  syn_burden_test=fisher.test(matrix(c(sum(cohort_syn$ac_case),num_case,  sum(cohort_syn$ac_ctrl), num_ctrl), nrow=2))
  
  odds_ratio_summary_ptv[i,]=c(sum(cohort_ptv$ac_case), sum(cohort_ptv$ac_ctrl), ptv_burden_test$estimate, ptv_burden_test$conf.int[1], ptv_burden_test$conf.int[2])
  odds_ratio_summary_syn[i,]=c(sum(cohort_syn$ac_case), sum(cohort_syn$ac_ctrl), syn_burden_test$estimate, syn_burden_test$conf.int[1], syn_burden_test$conf.int[2])
}
rownames(odds_ratio_summary_ptv)=cohorts
rownames(odds_ratio_summary_syn)=cohorts
```


## Replicate Extended Data Fig 2  

```{r, echo=F, message=F, warning=F}
#################### population groups
unique(SCHEMA_data$group)

############ convert into numeric values 
SCHEMA_data$ac_case=as.numeric(SCHEMA_data$ac_case)
SCHEMA_data$ac_ctrl=as.numeric(SCHEMA_data$ac_ctrl)

############ minor allele count filter 
MAC_threshold=5 # this cutoff is used in SCHEMA paper in Section Analysis approach  
SCHEMA_syn=SCHEMA_data %>% filter(group =="meta" & consequence=="synonymous_variant" & ac_case<=MAC_threshold & ac_ctrl<=MAC_threshold) 

##############
# what are ac_case, ac_ctrl, an_case, an_ctrl? 
# ac_case: allle count in case; ac_ctrl: allele count in control; an_case: allele number in case; an_ctrl: allele number in control #  see details here: https://schema.broadinstitute.org/gene/ENSG00000055130; move cursor to the variables then their meanings will pop up  


############## check allele frequency 
#max(SCHEMA_syn$ac_case, na.rm=T)
#max(SCHEMA_syn$ac_ctrl, na.rm=T)
#SCHEMA_syn %>% filter(ac_case=="48587") from here, SCHEMA data doesn't use AF filter. 

############### overall Burden of syn mutations 
#SCHEMA_syn$ac_case=as.numeric(SCHEMA_syn$ac_case) # convert into numeric value 
#SCHEMA_syn$ac_ctrl=as.numeric(SCHEMA_syn$ac_ctrl)
#SCHEMA_syn=SCHEMA_syn[complete.cases(SCHEMA_syn$ac_case),] # remove rows when either ac_case or ac_ctrl has NA 
#SCHEMA_syn=SCHEMA_syn[complete.cases(SCHEMA_syn$ac_ctrl),]

num_variant=nrow(SCHEMA_syn)
ac_case_total=sum(SCHEMA_syn$ac_case)
ac_ctrl_total=sum(SCHEMA_syn$ac_ctrl)
#num_case=24248; num_ctrl=46885 # numbers from from Fig 1a https://www.nature.com/articles/s41586-022-04556-w#Sec15
#num_case=24248; num_ctrl=50437 # numbers from Table S2
#there are 50437 internal control and 46885 gnomAD controls   
num_case=22444; num_ctrl=39837  # mumbers from Fig 1c, from the same population, see Section-Statistical approach in suppl 

############################ 
OR=numeric(); OR_lower=numeric(); OR_upper=numeric()
#we considered frameshift, stop gained, splice acceptor
#and donor variants as putative protein-truncating (or loss-of-function) variants, so LoF and PTV have the same definition. 
LoF=SCHEMA_data %>% filter(consequence=="frameshift_variant" | consequence=="stop_gained" |consequence=="splice_acceptor_variant" | consequence=="splice_donor_variant")
ac_case_total=sum(as.numeric(LoF$ac_case), na.rm=T); ac_ctrl_total=sum(as.numeric(LoF$ac_ctrl), na.rm=T)
burden_test=fisher.test(matrix(c(ac_case_total,num_case, ac_ctrl_total, num_ctrl), nrow=2))
OR[1]=burden_test$estimate; OR_lower[1]=burden_test$conf.int[1]; OR_upper[1]=burden_test$conf.int[2]


```

## gene CUL1

```{r, echo=F, message=F, warning=F}
CUL1=SCHEMA_data %>% filter(gene_id=="ENSG00000055130" & consequence!="synonymous_variant") # https://schema.broadinstitute.org/gene/ENSG00000055130

#CUL1 %>% filter(locus=="7:148427164") # zoom in this single variant here https://schema.broadinstitute.org/gene/ENSG00000055130
#to know:  (genome): AFR(genomes); (exomes): EUR(exomes); the third one is meta; 
```

