---
title: "SCHEMA_genomewide"
output: html_document
date: "2023-11-09"
---

```{r, echo=F, warning=F, message=F}
rm(list=ls())
set.seed(123)
knitr::opts_chunk$set(autodep = TRUE)
library(knitr)
library(RSQLite)
library(dplyr)
library(knitr)
knitr::opts_chunk$set(cache = TRUE, warning = FALSE, 
                      message = FALSE, cache.lazy = FALSE) # avoid error of long vectors https://stackoverflow.com/questions/39417003/long-vectors-not-supported-yet-error-in-rmd-but-not-in-r-script 
library(kableExtra)
library(RColorBrewer)
library(gplots)
library(tidyverse)
library(gridExtra)
library(ggpubr)
library(DT)
#library("devtools")
#install.packages("Rtools")
#install_github('xinhe-lab/mirage')
library(mirage)

multiplesheets <- function(fname) {
   
  # getting info about all excel sheets
  sheets <- readxl::excel_sheets(fname)
  tibble <- lapply(sheets, function(x) readxl::read_excel(fname, sheet = x))
  data_frame <- lapply(tibble, as.data.frame)
    
  # assigning names to data frames
  names(data_frame) <- sheets
    
  # print data frame
  print(data_frame)
}

pvalue_adjust=function(p_value)
{
  p_value_adjust=numeric()
for (i in 1:length(p_value))
{
  if (is.na(p_value[i])==T)
    p_value_adjust[i]=p_value[i]
  if (is.na(p_value[i])==F & p_value[i]<1e-10)
    p_value_adjust[i]="<1e-10"
  if (is.na(p_value[i])==F & p_value[i]>1e-10)
    p_value_adjust[i]=round(p_value[i],10)
}
  return(p_value_adjust)
}

```




```{r, echo=F, message=F, warning=F, eval=F}
SCHEMA_data=as_tibble(read.table("C:\\han\\Dataset\\SCHEMA\\SCHEMA_variant_results_subset_20231025.tsv", header=T))
```




```{r, echo=F, message=F, warning=F, eval=F}
# Conversion of Gene Name to Ensembl ID https://www.biostars.org/p/430015/

# install package biomaRT
#if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

#BiocManager::install("biomaRt")

library(biomaRt)
mart <- useMart('ENSEMBL_MART_ENSEMBL')
mart <- useDataset('hsapiens_gene_ensembl', mart)

annotLookup <- getBM(
  mart = mart,
  attributes = c(
    'hgnc_symbol',
    'ensembl_gene_id',
    'gene_biotype'),
  uniqueRows = TRUE)
colnames(annotLookup)[2]="gene_id"

SCHEMA_data_with_hgnc_symbol=as_tibble(right_join(annotLookup, SCHEMA_data, by="gene_id")) 
```



## meta case control sample


### syn mutations only 

```{r, echo=F, message=F, warning=F, eval=F}
MAC_threshold=5
SCHEMA_data_with_hgnc_symbol$ac_case=as.numeric(SCHEMA_data_with_hgnc_symbol$ac_case)  # convert into numeric values 
SCHEMA_data_with_hgnc_symbol$ac_ctrl=as.numeric(SCHEMA_data_with_hgnc_symbol$ac_ctrl)

meta_data=SCHEMA_data_with_hgnc_symbol %>% filter(group=="meta" & ac_case<=MAC_threshold & ac_ctrl<=MAC_threshold & hgnc_symbol!="") %>% filter(consequence=="synonymous_variant" ) %>% dplyr::select(hgnc_symbol, locus, ac_case, ac_ctrl, n_denovos, consequence)
meta_data$n_denovos[is.na(meta_data$n_denovos)]=0 # replace NA with zero  
ac_case_actual=meta_data$ac_case-meta_data$n_denovos # this is actual number of allele counts in cases # see here https://schema.broadinstitute.org/gene/ENSG00000099381
ac_ctrl_actual=meta_data$ac_ctrl
group_index=numeric()
group_index[which(meta_data$consequence=="synonymous_variant")]=1

meta_data_mirage=data.frame(ID=meta_data$locus, gene=meta_data$hgnc_symbol, No.case=ac_case_actual, No.contr=ac_ctrl_actual, group.index=group_index)
meta_data_mirage=meta_data_mirage %>% filter(No.contr-No.case<2) # filter out variants with too many alleles in control than in case 
n1=24248; n2=97322 
res=mirage(meta_data_mirage, n1, n2, gamma=1)

######### given a gene, check the input data with SCHEMA browser https://schema.broadinstitute.org/gene/ENSG00000055130
#check_gene="TTN"
#sum(meta_data_mirage %>% filter(gene==check_gene & group.index==1)%>% dplyr::select(No.case)%>% pull())
#sum(meta_data_mirage %>% filter(gene==check_gene & group.index==1)%>% dplyr::select(No.contr)%>% pull())

########### save the mirage output 
#write.csv(res$BF.PP.gene, file="C:\\han\\Dataset\\SCHEMA\\MIRAGE_on_meta_sample_estimate_delta_3vargroups_20231027.csv")

#res$BF.PP.gene[order(res$BF.PP.gene$BF, decreasing = T),] # order genes by BF 
#meta_data_mirage %>% filter(gene=="TTN")
```


```{r, echo=F, message=F, warning=F}

p_value=c(1, 1)


mirage_para_est=data.frame(variant_groups=c("delta:proportion of risk genes", "syn" ), parameter_estimate=round(c(0.0003262156,  
0.001006147),4), p_value=pvalue_adjust(p_value))
mirage_para_est%>%
datatable(extensions = 'Buttons',
            options = list(dom = 'Blfrtip',
                           buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                           lengthMenu = list(c(10,25,50,-1),
                                             c(10,25,50,"All"))))

var_group=c("syn")
estimate=c(0.001006147)
summary=data.frame(var.group=var_group, estimate=estimate)
ggplot(summary, aes(x=var.group, y=estimate, fill=var.group)) + 
   geom_bar(stat="identity", position=position_dodge())+
  theme(legend.title=element_blank())+
  xlab("variant groups")+ylab(expression(paste(hat(eta))))+
  ylim(c(0,1))+
  ggtitle("")+
  theme(plot.title = element_text(hjust = 0.5, size=10))+  #center the title 
  theme(axis.text.x = element_text(angle = 30, vjust = 0.5, size=10))
```

 * use filter `No.contr-No.case<2` filtering out variants with more alleles in cases than in control 



### estimate both $\delta$ and $\eta's$




#### 3 variant groups 

```{r, echo=F, message=F, warning=F, eval=F}
PTV_definition=c("frameshift_variant", "stop_gained", "splice_acceptor_variant","splice_donor_variant" )

SCHEMA_data_with_hgnc_symbol$ac_case=as.numeric(SCHEMA_data_with_hgnc_symbol$ac_case)  # convert into numeric values 
SCHEMA_data_with_hgnc_symbol$ac_ctrl=as.numeric(SCHEMA_data_with_hgnc_symbol$ac_ctrl)

meta_data=SCHEMA_data_with_hgnc_symbol %>% filter(group=="meta" & ac_case<=MAC_threshold & ac_ctrl<=MAC_threshold & hgnc_symbol!="") %>% filter(consequence %in% PTV_definition==T | consequence=="missense_variant_mpc_>=3" | consequence=="missense_variant_mpc_2-3") %>% dplyr::select(hgnc_symbol, locus, ac_case, ac_ctrl, n_denovos, consequence)
meta_data$n_denovos[is.na(meta_data$n_denovos)]=0 # replace NA with zero  
ac_case_actual=meta_data$ac_case-meta_data$n_denovos # this is actual number of allele counts in cases # see here https://schema.broadinstitute.org/gene/ENSG00000099381
ac_ctrl_actual=meta_data$ac_ctrl
group_index=numeric()
group_index[which(meta_data$consequence %in% PTV_definition==T)]=1
group_index[which(meta_data$consequence=="missense_variant_mpc_>=3")]=2
group_index[which(meta_data$consequence=="missense_variant_mpc_2-3")]=3
meta_data_mirage=data.frame(ID=meta_data$locus, gene=meta_data$hgnc_symbol, No.case=ac_case_actual, No.contr=ac_ctrl_actual, group.index=group_index)
n1=24248; n2=97322 
res=mirage(meta_data_mirage, n1, n2, gamma=c(6,3,1))

######### given a gene, check the input data with SCHEMA browser https://schema.broadinstitute.org/gene/ENSG00000055130
#check_gene="XPO7"
#sum(meta_data_mirage %>% filter(gene==check_gene & group.index==3)%>% dplyr::select(No.case)%>% pull())
#sum(meta_data_mirage %>% filter(gene==check_gene & group.index==3)%>% dplyr::select(No.contr)%>% pull())

########### save the mirage output 
#write.csv(res$BF.PP.gene, file="C:\\han\\Dataset\\SCHEMA\\MIRAGE_on_meta_sample_estimate_delta_3vargroups_20231027.csv")

```


```{r, echo=F, message=F, warning=F}


p_value=c(0, 0.000000e+00, 2.030313e-10, 1.000000e+00)


mirage_para_est=data.frame(variant_groups=c("delta:proportion of risk genes","PTV", "missense_variant_mpc_>=3", "missense_variant_mpc_2-3"), parameter_estimate=round(c(0.5606701,  
0.10127670, 0.25945529, 0.03010527),4), p_value=pvalue_adjust(p_value))
mirage_para_est%>%
datatable(extensions = 'Buttons',
            options = list(dom = 'Blfrtip',
                           buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                           lengthMenu = list(c(10,25,50,-1),
                                             c(10,25,50,"All"))))

var_group=c("(1) PTV", "(2) missense_variant_mpc_>=3", "(3) missense_variant_mpc_2-3")
estimate=c(0.10127670, 0.25945529, 0.03010527)
summary=data.frame(var.group=var_group, estimate=estimate)
ggplot(summary, aes(x=var.group, y=estimate, fill=var.group)) + 
   geom_bar(stat="identity", position=position_dodge())+
  theme(legend.title=element_blank())+
  xlab("variant groups")+ylab(expression(paste(hat(eta))))+
  ylim(c(0,1))+
  ggtitle("")+
  theme(plot.title = element_text(hjust = 0.5, size=10))+  #center the title 
  theme(axis.text.x = element_text(angle = 30, vjust = 0.5, size=10))
```


#### 2 variant groups 



```{r, echo=F, message=F, warning=F, eval=F}
MAC_threshold=5
PTV_definition=c("frameshift_variant", "stop_gained", "splice_acceptor_variant","splice_donor_variant" )
SCHEMA_data_with_hgnc_symbol$ac_case=as.numeric(SCHEMA_data_with_hgnc_symbol$ac_case)  # convert into numeric values 
SCHEMA_data_with_hgnc_symbol$ac_ctrl=as.numeric(SCHEMA_data_with_hgnc_symbol$ac_ctrl)

meta_data=SCHEMA_data_with_hgnc_symbol %>% filter(group=="meta" & ac_case<=MAC_threshold & ac_ctrl<=MAC_threshold & hgnc_symbol!="") %>% filter(consequence %in% PTV_definition==T | consequence=="missense_variant_mpc_>=3" | consequence=="missense_variant_mpc_2-3") %>% dplyr::select(hgnc_symbol, locus, ac_case, ac_ctrl, n_denovos, consequence)
meta_data$n_denovos[is.na(meta_data$n_denovos)]=0 # replace NA with zero  
ac_case_actual=meta_data$ac_case-meta_data$n_denovos # this is actual number of allele counts in cases # see here https://schema.broadinstitute.org/gene/ENSG00000099381
ac_ctrl_actual=meta_data$ac_ctrl
group_index=numeric()
group_index[which(meta_data$consequence %in% PTV_definition==T | meta_data$consequence=="missense_variant_mpc_>=3")]=1
group_index[which(meta_data$consequence=="missense_variant_mpc_2-3")]=2
meta_data_mirage=data.frame(ID=meta_data$locus, gene=meta_data$hgnc_symbol, No.case=ac_case_actual, No.contr=ac_ctrl_actual, group.index=group_index)
n1=24248; n2=97322 
res=mirage(meta_data_mirage, n1, n2, gamma=c(3,1))

######### given a gene, check the input data with SCHEMA browser https://schema.broadinstitute.org/gene/ENSG00000055130
#check_gene="SETD1A"
#sum(meta_data_mirage %>% filter(gene==check_gene & group.index==1)%>% dplyr::select(No.case)%>% pull())
#sum(meta_data_mirage %>% filter(gene==check_gene & group.index==1)%>% dplyr::select(No.contr)%>% pull())

########### save the mirage output 
#write.csv(res$BF.PP.gene, file="C:\\han\\Dataset\\SCHEMA\\MIRAGE_on_meta_sample_estimate_delta_2vargroups_20231027.csv")
#readr::write_rds(res$BF.all, "C:\\han\\Dataset\\SCHEMA\\MIRAGE_on_meta_sample_estimate_delta_2vargroups_BF_all_20231110.Rds") # save large dataset in R 
BF_all=readr::read_rds("C:\\han\\Dataset\\SCHEMA\\MIRAGE_on_meta_sample_estimate_delta_2vargroups_BF_all_20231110.Rds")
BF.PP.gene=as_tibble(read.csv("C:\\han\\Dataset\\SCHEMA\\MIRAGE_on_meta_sample_estimate_delta_2vargroups_20231027.csv", header=T))
######## check top genes and variants driving the signal 
BF.PP.gene[order(BF.PP.gene$BF, decreasing = T),]
BF_all[[1265]]
sum(BF_all[[15842]]$No.case)
sum(BF_all[[15842]]$No.contr)

```





```{r, echo=F, message=F, warning=F}

p_value=c(0, 0.000000e+00, 1.000000e+00)


mirage_para_est=data.frame(variant_groups=c("delta:proportion of risk genes","PTV or missense_variant_mpc_>=3", "missense_variant_mpc_2-3"), parameter_estimate=round(c(0.5747217,  
0.19136120, 0.03100808),4), p_value=pvalue_adjust(p_value))
mirage_para_est%>%
datatable(extensions = 'Buttons',
            options = list(dom = 'Blfrtip',
                           buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                           lengthMenu = list(c(10,25,50,-1),
                                             c(10,25,50,"All"))))

var_group=c("(1) PTV or missense_variant_mpc_>=3", "(2) missense_variant_mpc_2-3")
estimate=c(0.19136120, 0.03100808)
summary=data.frame(var.group=var_group, estimate=estimate)
ggplot(summary, aes(x=var.group, y=estimate, fill=var.group)) + 
   geom_bar(stat="identity", position=position_dodge())+
  theme(legend.title=element_blank())+
  xlab("variant groups")+ylab(expression(paste(hat(eta))))+
  ylim(c(0,1))+
  ggtitle("")+
  theme(plot.title = element_text(hjust = 0.5, size=10))+  #center the title 
  theme(axis.text.x = element_text(angle = 30, vjust = 0.5, size=10))
```




#### Bayesian FDR


```{r, echo=F, message=F, warning=F, eval=T}

##############
########## this is the result with parameteres fixed at 5% 
BF.PP.gene=as_tibble(read.csv("C:\\han\\Dataset\\SCHEMA\\MIRAGE_on_meta_sample_estimate_delta_2vargroups_20231027.csv", header=T))


fdr_function=function(fdr_threshold, data)
{
tau=seq(0, 0.999, by=0.001)
num_pred=NULL
false_disc=NULL
FDR=NULL
for (i in 1:length(tau))
{
num_pred[i]=sum(ifelse(data$post_prob>tau[i], 1, 0))
false_disc[i]=sum((1-data$post_prob)*ifelse(data$post_prob>tau[i], 1, 0))
FDR[i]=false_disc[i]/num_pred[i]
}
tau_fdr=tibble(tau=tau, fdr=FDR)%>%drop_na()  # drop rows with NA
tau_at_threshold=tau_fdr%>%filter(fdr<=fdr_threshold)
if (nrow(tau_at_threshold)==0)
  return("No gene is selected")
if (nrow(tau_at_threshold)>0)
{
  tau_threshold=tau_at_threshold$tau[1]
  return(data%>% filter(post_prob>tau_threshold))
}

}
```

```{r, echo=T, message=F, warning=F, eval=T}
input_data=BF.PP.gene
colnames(input_data)[4]="post_prob"
FDR_gene=fdr_function(fdr_threshold = 0.05, input_data)
```


```{r, echo=F, message=F, warning=F, results=F}
SCHEMA_gene_results=multiplesheets("C:\\han\\Dataset\\SCHEMA\\41586_2022_4556_MOESM3_ESM.xlsx")
```

```{r, echo=F, message=F, warning=F}
SCHEMA_top_gene=SCHEMA_gene_results$`Table S5 - Gene Results`$`Gene Symbol`[1:10]
SCHEMA_top_gene2=SCHEMA_gene_results$`Table S5 - Gene Results`[order(SCHEMA_gene_results$`Table S5 - Gene Results`$`P ca/co (comb)`, decreasing = F),]$`Gene Symbol`[1:10]
library("ggvenn")
x=list(
  fdr_gene=FDR_gene$Gene,
  scheme_gene_with_denovo=SCHEMA_top_gene
)
ggvenn(x)+
  ggtitle("")+
  theme(plot.title = element_text(hjust = 0.5, size=20, color = 2)) #center the title 



x=list(
  fdr_gene=FDR_gene$Gene,
  scheme_gene_without_denovo=SCHEMA_top_gene2
)
ggvenn(x)+
  ggtitle("")+
  theme(plot.title = element_text(hjust = 0.5, size=20, color = 2)) #center the title

```