---
title: "SCHEMA_constraint_gene"
output: html_document
date: "2023-11-15"
---

```{r, echo=F, warning=F, message=F}
rm(list=ls())
set.seed(123)
knitr::opts_chunk$set(autodep = TRUE)
library(knitr)
library(RSQLite)
library(dplyr)
library(knitr)
knitr::opts_chunk$set(cache = TRUE, warning = FALSE, 
                      message = FALSE, cache.lazy = FALSE) # avoid error of long vectors https://stackoverflow.com/questions/39417003/long-vectors-not-supported-yet-error-in-rmd-but-not-in-r-script 
library(kableExtra)
library(RColorBrewer)
library(gplots)
library(tidyverse)
library(gridExtra)
library(ggpubr)
library(DT)
#library("devtools")
#install.packages("Rtools")
#install_github('xinhe-lab/mirage')
library(mirage)

multiplesheets <- function(fname) {
   
  # getting info about all excel sheets
  sheets <- readxl::excel_sheets(fname)
  tibble <- lapply(sheets, function(x) readxl::read_excel(fname, sheet = x))
  data_frame <- lapply(tibble, as.data.frame)
    
  # assigning names to data frames
  names(data_frame) <- sheets
    
  # print data frame
  print(data_frame)
}

pvalue_adjust=function(p_value)
{
  p_value_adjust=numeric()
for (i in 1:length(p_value))
{
  if (is.na(p_value[i])==T)
    p_value_adjust[i]=p_value[i]
  if (is.na(p_value[i])==F & p_value[i]<1e-10)
    p_value_adjust[i]="<1e-10"
  if (is.na(p_value[i])==F & p_value[i]>1e-10)
    p_value_adjust[i]=round(p_value[i],10)
}
  return(p_value_adjust)
}

```




```{r, echo=F, message=F, warning=F, eval=F}
SCHEMA_data=as_tibble(read.table("C:\\han\\Dataset\\SCHEMA\\SCHEMA_variant_results_subset_20231025.tsv", header=T))
```


```{r, echo=F, message=F, warning=F, results=F}
# get constraint genes 
Table13=multiplesheets("C:/han/Dataset/20231017_PMID_27535533/Lek 2016 suppl table/nature19057-s2/nature19057-SI Table 13.xlsx")


#suppl_table11=read.table(gzfile("../../../../Dataset/20231018_PMID_32461654/41586_2020_2308_MOESM4_ESM/supplement/supplementary_dataset_11_full_constraint_metrics.tsv.gz"),fill = TRUE , header = TRUE ) # read into gz file 
# length(unique(suppl_table11 %>% filter(pLI>0.9) %>% select(gene) %>%pull)) 3476 genes with pLI>0.9 
suppl_table11=read.table(gzfile("C:/han/Dataset/20231018_PMID_32461654/41586_2020_2308_MOESM4_ESM/supplement/supplementary_dataset_11_full_constraint_metrics.tsv.gz"),fill = TRUE , header = TRUE ) # read into gz file
```




```{r, echo=F, message=F, warning=F, eval=F}
# Conversion of Gene Name to Ensembl ID https://www.biostars.org/p/430015/

# install package biomaRT
#if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

#BiocManager::install("biomaRt")

library(biomaRt)
mart <- useMart('ENSEMBL_MART_ENSEMBL')
mart <- useDataset('hsapiens_gene_ensembl', mart)

annotLookup <- getBM(
  mart = mart,
  attributes = c(
    'hgnc_symbol',
    'ensembl_gene_id',
    'gene_biotype'),
  uniqueRows = TRUE)
colnames(annotLookup)[2]="gene_id"

SCHEMA_data_with_hgnc_symbol=as_tibble(right_join(annotLookup, SCHEMA_data, by="gene_id")) 
```

## meta case control sample


### estimate both $\delta$ and $\eta's$




#### 3 variant groups 

```{r, echo=F, message=F, warning=F, eval=F}
############ get constraint genes with pLI>0.9  
LoF_intolerant=Table13$`LoF Intolerant`
# length(unique(LoF_intolerant %>% filter(pLI>0.9) %>% (gene) %>% pull)) # 3230 genes with pLI>0.9, consistent with the number reported in paper Lek2017_Nature 

LoF_intolerant_gene=unique(suppl_table11 %>% filter(pLI>0.9) %>% dplyr::select(gene) %>%pull) 



PTV_definition=c("frameshift_variant", "stop_gained", "splice_acceptor_variant","splice_donor_variant" )

MAC_threshold=5

SCHEMA_data_with_hgnc_symbol$ac_case=as.numeric(SCHEMA_data_with_hgnc_symbol$ac_case)  # convert into numeric values 
SCHEMA_data_with_hgnc_symbol$ac_ctrl=as.numeric(SCHEMA_data_with_hgnc_symbol$ac_ctrl)

meta_data=SCHEMA_data_with_hgnc_symbol %>% filter(hgnc_symbol %in% LoF_intolerant_gene==T & group=="meta" & ac_case<=MAC_threshold & ac_ctrl<=MAC_threshold & hgnc_symbol!="") %>% filter(consequence %in% PTV_definition==T | consequence=="missense_variant_mpc_>=3" | consequence=="missense_variant_mpc_2-3") %>% dplyr::select(hgnc_symbol, locus, ac_case, ac_ctrl, n_denovos, consequence)
meta_data$n_denovos[is.na(meta_data$n_denovos)]=0 # replace NA with zero  
ac_case_actual=meta_data$ac_case-meta_data$n_denovos # this is actual number of allele counts in cases # see here https://schema.broadinstitute.org/gene/ENSG00000099381
ac_ctrl_actual=meta_data$ac_ctrl
group_index=numeric()
group_index[which(meta_data$consequence %in% PTV_definition==T)]=1
group_index[which(meta_data$consequence=="missense_variant_mpc_>=3")]=2
group_index[which(meta_data$consequence=="missense_variant_mpc_2-3")]=3
meta_data_mirage=data.frame(ID=meta_data$locus, gene=meta_data$hgnc_symbol, No.case=ac_case_actual, No.contr=ac_ctrl_actual, group.index=group_index)
n1=24248; n2=97322 
res=mirage(meta_data_mirage, n1, n2, gamma=c(6,3,1))

######### given a gene, check the input data with SCHEMA browser https://schema.broadinstitute.org/gene/ENSG00000055130
#check_gene="XPO7"
#sum(meta_data_mirage %>% filter(gene==check_gene & group.index==3)%>% dplyr::select(No.case)%>% pull())
#sum(meta_data_mirage %>% filter(gene==check_gene & group.index==3)%>% dplyr::select(No.contr)%>% pull())

########### save the mirage output 
#write.csv(res$BF.PP.gene, file="C:\\han\\Dataset\\SCHEMA\\MIRAGE_on_meta_sample_estimate_delta_3vargroups_20231027.csv")
# first extract, then save to RData
BF.PP.gene=res$BF.PP.gene
delta.est=res$delta.est 
delta.pvalue=res$delta.pvalue
eta.est=res$eta.est
eta.pvalue=res$eta.pvalue
BF.all=res$BF.all
#save(BF.PP.gene, delta.est, delta.pvalue, eta.est, eta.pvalue, BF.all, file="C:\\han\\Dataset\\SCHEMA\\MIRAGE_on_meta_sample_estimate_delta_3vargroups_for_constraint_gene_20231115.RData")
```


```{r, echo=F, message=F, warning=F}


p_value=c(5.277693e-216,  2.116778e-209,  1.938024e-12, 1.000000e+00)


mirage_para_est=data.frame(variant_groups=c("delta:proportion of risk genes","PTV", "missense_variant_mpc_>=3", "missense_variant_mpc_2-3"), parameter_estimate=round(c(0.4477384,  
0.384832987, 0.519311643, 0.008214116),4), p_value=pvalue_adjust(p_value))
mirage_para_est%>%
datatable(extensions = 'Buttons',
            options = list(dom = 'Blfrtip',
                           buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                           lengthMenu = list(c(10,25,50,-1),
                                             c(10,25,50,"All"))))

var_group=c("(1) PTV", "(2) missense_variant_mpc_>=3", "(3) missense_variant_mpc_2-3")
estimate=c(0.384832987, 0.519311643, 0.008214116)
summary=data.frame(var.group=var_group, estimate=estimate)
ggplot(summary, aes(x=var.group, y=estimate, fill=var.group)) + 
   geom_bar(stat="identity", position=position_dodge())+
  theme(legend.title=element_blank())+
  xlab("")+ylab(expression(paste(hat(eta))))+
  ylim(c(0,1))+
  ggtitle("")+
  theme(plot.title = element_text(hjust = 0.5, size=10))+  #center the title 
  theme(axis.text.x = element_text(angle = 30, vjust = 0.5, size=10))

```





#### Bayesian FDR


```{r, echo=F, message=F, warning=F, eval=T}

##############
##########
#BF.PP.gene=as_tibble(read.csv("C:\\han\\Dataset\\SCHEMA\\MIRAGE_on_meta_sample_estimate_delta_3vargroups_20231027.csv", header=T))
load("C:\\han\\Dataset\\SCHEMA\\MIRAGE_on_meta_sample_estimate_delta_3vargroups_for_constraint_gene_20231115.RData")

fdr_function=function(fdr_threshold, data)
{
tau=seq(0, 0.999, by=0.001)
num_pred=NULL
false_disc=NULL
FDR=NULL
for (i in 1:length(tau))
{
num_pred[i]=sum(ifelse(data$post_prob>tau[i], 1, 0))
false_disc[i]=sum((1-data$post_prob)*ifelse(data$post_prob>tau[i], 1, 0))
FDR[i]=false_disc[i]/num_pred[i]
}
tau_fdr=tibble(tau=tau, fdr=FDR)%>%drop_na()  # drop rows with NA
tau_at_threshold=tau_fdr%>%filter(fdr<=fdr_threshold)
if (nrow(tau_at_threshold)==0)
  return("No gene is selected")
if (nrow(tau_at_threshold)>0)
{
  tau_threshold=tau_at_threshold$tau[1]
  return(data%>% filter(post_prob>tau_threshold))
}

}
```



```{r, echo=T, message=F, warning=F, eval=T}
input_data=BF.PP.gene
colnames(input_data)[3]="post_prob"
FDR_gene=fdr_function(fdr_threshold = 0.05, input_data)
```


```{r, echo=F, message=F, warning=F, results=F}
SCHEMA_gene_results=multiplesheets("C:\\han\\Dataset\\SCHEMA\\41586_2022_4556_MOESM3_ESM.xlsx")
```


* overlap with top 10 exome significant genes 


```{r, echo=F, message=F, warning=F}
SCHEMA_top_gene=SCHEMA_gene_results$`Table S5 - Gene Results`$`Gene Symbol`[1:10]
SCHEMA_top_gene2=SCHEMA_gene_results$`Table S5 - Gene Results`[order(SCHEMA_gene_results$`Table S5 - Gene Results`$`P ca/co (comb)`, decreasing = F),]$`Gene Symbol`[1:10]
library("ggvenn")
x=list(
  fdr_gene=FDR_gene$Gene,
  scheme_gene_with_denovo=SCHEMA_top_gene
)
#ggvenn(x)+
#  ggtitle("")+
#  theme(plot.title = element_text(hjust = 0.5, size=20, color = 2)) #center the title 



x=list(
  fdr_gene=FDR_gene$Gene,
  scheme_gene_without_denovo=SCHEMA_top_gene2
)
#ggvenn(x)+
#  ggtitle("")+
#  theme(plot.title = element_text(hjust = 0.5, size=20, color = 2)) #center the title

```




* overlap with top 33 genes at FDR 5\% 


```{r, echo=F, message=F, warning=F}
SCHEMA_top_gene=SCHEMA_gene_results$`Table S5 - Gene Results`$`Gene Symbol`[1:33]
SCHEMA_top_gene2=SCHEMA_gene_results$`Table S5 - Gene Results`[order(SCHEMA_gene_results$`Table S5 - Gene Results`$`P ca/co (comb)`, decreasing = F),]$`Gene Symbol`[1:33]
library("ggvenn")
x=list(
  fdr_gene=FDR_gene$Gene,
  scheme_gene_with_denovo=SCHEMA_top_gene
)
#ggvenn(x)+
#  ggtitle("")+
#  theme(plot.title = element_text(hjust = 0.5, size=20, color = 2)) #center the title 



x=list(
  fdr_gene=FDR_gene$Gene,
  scheme_gene_without_denovo=SCHEMA_top_gene2
)
#ggvenn(x)+
#  ggtitle("")+
#  theme(plot.title = element_text(hjust = 0.5, size=20, color = 2)) #center the title

```




#### 2 variant groups 

```{r, echo=F, message=F, warning=F, eval=F}
############ get constraint genes with pLI>0.9  
LoF_intolerant=Table13$`LoF Intolerant`
# length(unique(LoF_intolerant %>% filter(pLI>0.9) %>% (gene) %>% pull)) # 3230 genes with pLI>0.9, consistent with the number reported in paper Lek2017_Nature 

LoF_intolerant_gene=unique(suppl_table11 %>% filter(pLI>0.9) %>% dplyr::select(gene) %>%pull) 



PTV_definition=c("frameshift_variant", "stop_gained", "splice_acceptor_variant","splice_donor_variant" )

MAC_threshold=5

SCHEMA_data_with_hgnc_symbol$ac_case=as.numeric(SCHEMA_data_with_hgnc_symbol$ac_case)  # convert into numeric values 
SCHEMA_data_with_hgnc_symbol$ac_ctrl=as.numeric(SCHEMA_data_with_hgnc_symbol$ac_ctrl)

meta_data=SCHEMA_data_with_hgnc_symbol %>% filter(hgnc_symbol %in% LoF_intolerant_gene==T & group=="meta" & ac_case<=MAC_threshold & ac_ctrl<=MAC_threshold & hgnc_symbol!="") %>% filter(consequence %in% PTV_definition==T | consequence=="missense_variant_mpc_>=3" | consequence=="missense_variant_mpc_2-3") %>% dplyr::select(hgnc_symbol, locus, ac_case, ac_ctrl, n_denovos, consequence)
meta_data$n_denovos[is.na(meta_data$n_denovos)]=0 # replace NA with zero  
ac_case_actual=meta_data$ac_case-meta_data$n_denovos # this is actual number of allele counts in cases # see here https://schema.broadinstitute.org/gene/ENSG00000099381
ac_ctrl_actual=meta_data$ac_ctrl
group_index=numeric()
group_index[which(meta_data$consequence %in% PTV_definition==T | meta_data$consequence=="missense_variant_mpc_>=3")]=1
group_index[which(meta_data$consequence=="missense_variant_mpc_2-3")]=2
meta_data_mirage=data.frame(ID=meta_data$locus, gene=meta_data$hgnc_symbol, No.case=ac_case_actual, No.contr=ac_ctrl_actual, group.index=group_index)
n1=24248; n2=97322 
res=mirage(meta_data_mirage, n1, n2, gamma=c(3,1))

######### given a gene, check the input data with SCHEMA browser https://schema.broadinstitute.org/gene/ENSG00000055130
#check_gene="XPO7"
#sum(meta_data_mirage %>% filter(gene==check_gene & group.index==3)%>% dplyr::select(No.case)%>% pull())
#sum(meta_data_mirage %>% filter(gene==check_gene & group.index==3)%>% dplyr::select(No.contr)%>% pull())

########### save the mirage output 
#write.csv(res$BF.PP.gene, file="C:\\han\\Dataset\\SCHEMA\\MIRAGE_on_meta_sample_estimate_delta_3vargroups_20231027.csv")
# first extract, then save to RData
BF.PP.gene=res$BF.PP.gene
delta.est=res$delta.est 
delta.pvalue=res$delta.pvalue
eta.est=res$eta.est
eta.pvalue=res$eta.pvalue
BF.all=res$BF.all
#save(BF.PP.gene, delta.est, delta.pvalue, eta.est, eta.pvalue, BF.all, #file="C:\\han\\Dataset\\SCHEMA\\MIRAGE_on_meta_sample_estimate_delta_2vargroups_for_constraint_gene_20231120.RData")
```



```{r, echo=F, message=F, warning=F}

p_value=c(1.035716e-213, 8.583023e-217,  1.000000e+00)

mirage_para_est_2group=data.frame(variant_groups=c("delta:proportion of risk genes","PTV or missense_variant_mpc_>=3", "missense_variant_mpc_2-3"), parameter_estimate=round(c(0.4507106,  
0.676594081, 0.008126521),4), p_value=pvalue_adjust(p_value))
mirage_para_est_2group%>%
datatable(extensions = 'Buttons',
            options = list(dom = 'Blfrtip',
                           buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                           lengthMenu = list(c(10,25,50,-1),
                                             c(10,25,50,"All"))))

var_group=c("(1) PTV or missense variant mpc_>=3", "(2) missense variant mpc_2-3")
estimate=c(0.676594081, 0.008126521)
summary=data.frame(var.group=var_group, estimate=estimate)

#png("C:\\Users\\hans\\OneDrive - Marquette University\\SCHEMA\\Fig_parameter_estimate.png")

ggplot(summary, aes(x=var.group, y=estimate, fill=var.group)) + 
   geom_bar(stat="identity", position=position_dodge())+
  theme(legend.title=element_blank())+
  xlab("")+ylab(expression(paste(hat(eta))))+
  ylim(c(0,1))+
  ggtitle(" Estimated proportion of risk variants in variant groups")+
  theme(plot.title = element_text(hjust = 0.5, size=15))+  #center the title 
  theme(axis.text.x = element_text(angle = 0, vjust = 0.5, size=12))+
  theme(legend.position="none")
#dev.off()
```


#### hyper parameters sensitivity 

```{r, echo=F, message=F, warning=F, eval=F}
# recover input data to run MIRAGE 
load("C:\\han\\Dataset\\SCHEMA\\MIRAGE_on_meta_sample_estimate_delta_2vargroups_for_constraint_gene_20231120.RData")
input1=do.call(rbind.data.frame, BF.all) 
input=input1[,1:5]
n1=24248; n2=97322 
gamma_bar=c(2,3,4,5)
parameter_estimate=matrix(nrow=length(gamma_bar), ncol=3)
parameter_estimate_pvalue=matrix(nrow=length(gamma_bar), ncol=3)

for (i in 1:length(gamma_bar))
{
cat(i, "th parameter is running", "\n")
res=mirage(input, n1, n2, gamma=c(gamma_bar[i],1))
parameter_estimate[i,]=c(res$delta.est, res$eta.est)
parameter_estimate_pvalue[i,]=c(res$delta.pvalue, res$eta.pvalue)
}
```

```{r, echo=F, message=F, warning=F}
library(scales)
parameter_estimate_plot=matrix(nrow=3, ncol=3)
parameter_estimate_plot[1,]=c(0.4507106, 0.6765941, 0.008126521) # pvalue: 1.035716e-213 8.583023e-217    1
parameter_estimate_plot[2,]=c(0.4515301, 0.5060380, 0.008264481) # pvalue: 2.074614e-215 1.702703e-218    1
parameter_estimate_plot[3,]=c(0.4514804, 0.4229823, 0.008357832) # pvalue: 3.107830e-216 2.539093e-219    1


rownames(parameter_estimate_plot)=c("gamma_bar=3", "gamma_bar=4", "gamma_bar=5")
colnames(parameter_estimate_plot)=c("delta", "eta1", "eta2")

data_plot2=data.frame(parameters_estimate=as.vector(parameter_estimate_plot), parameters=rep(c("delta", "eta1", "eta2"), each=3), gamma_bar=rep(c("3", "4", "5"),3))

data_plot= data_plot2 %>%
  mutate(parameters = recode_factor(parameters, `delta` = "delta", `eta1` = "eta[1]", `eta2`="eta[2]"))


#png("C:\\Users\\hans\\OneDrive - Marquette University\\SCHEMA\\Suppl\\Fig_para_sensitivity.png")

point.size=3
text.size=15
ggplot(data_plot, aes(x=gamma_bar, y=parameters_estimate, group=parameters))+
  geom_line(aes(color=parameters), size=1)+
  geom_point(aes(color=parameters), size=point.size)+
  xlab(expression(paste(bar(gamma))))+ 
  ylab("Parameter Estimate")+
  ylim(c(0,1))+
  #ggtitle("Sensitivity analysis of hyper parameter")+
  labs(title = bquote("Sensitivity of " ~ bar(gamma)))+
  theme_classic()+
  theme(plot.title = element_text(hjust = 0.5, size=text.size))+
theme(text = element_text(size=text.size))+
 # theme(legend.position = "none")+
  scale_colour_discrete(labels = parse_format())
#dev.off()
```



#### Bayesian FDR


```{r, echo=F, message=F, warning=F, eval=T}

##############
##########
rm("BF.PP.gene")
#BF.PP.gene=as_tibble(read.csv("C:\\han\\Dataset\\SCHEMA\\MIRAGE_on_meta_sample_estimate_delta_3vargroups_20231027.csv", header=T))
load("C:\\han\\Dataset\\SCHEMA\\MIRAGE_on_meta_sample_estimate_delta_2vargroups_for_constraint_gene_20231120.RData")

fdr_function=function(fdr_threshold, data)
{
tau=seq(0, 0.999, by=0.001)
num_pred=NULL
false_disc=NULL
FDR=NULL
for (i in 1:length(tau))
{
num_pred[i]=sum(ifelse(data$post_prob>tau[i], 1, 0))
false_disc[i]=sum((1-data$post_prob)*ifelse(data$post_prob>tau[i], 1, 0))
FDR[i]=false_disc[i]/num_pred[i]
}
tau_fdr=tibble(tau=tau, fdr=FDR)%>%drop_na()  # drop rows with NA
tau_at_threshold=tau_fdr%>%filter(fdr<=fdr_threshold)
if (nrow(tau_at_threshold)==0)
  return("No gene is selected")
if (nrow(tau_at_threshold)>0)
{
  tau_threshold=tau_at_threshold$tau[1]
  return(data%>% filter(post_prob>tau_threshold))
}

}
```

```{r, echo=T, message=F, warning=F, eval=T}
input_data=BF.PP.gene
colnames(input_data)[3]="post_prob"
FDR_gene=fdr_function(fdr_threshold = 0.05, input_data)
```




#### BF in every variant group


```{r, echo=F, message=F, warning=F}
rm("BF.PP.gene")
#BF.PP.gene=as_tibble(read.csv("C:\\han\\Dataset\\SCHEMA\\MIRAGE_on_meta_sample_estimate_delta_3vargroups_20231027.csv", header=T))
load("C:\\han\\Dataset\\SCHEMA\\MIRAGE_on_meta_sample_estimate_delta_2vargroups_for_constraint_gene_20231120.RData")  # load MIRAGE results

num_category=2 # number of variant groups 
BF_variant_group=matrix(nrow=length(BF.all), ncol=num_category)
for (i in 1:length(BF.all))
{
variant_full_infor=BF.all[[i]]  # given a gene, say the first gene 
variant_group_BF=numeric() # store variant group BF for a given gene 
for (j in 1:num_category)
{
  if (sum(variant_full_infor$original.index==j)==0) # no variants in this group
    variant_group_BF[j]=NA
  
  if (sum(variant_full_infor$original.index==j)>0) # variants exist in this group
    variant_group_BF[j]=prod(1-eta.est[j]+eta.est[j]*
    variant_full_infor[which(variant_full_infor$original.index==j),]$var.BF) 
  # product of BF of variants within that variant group 
   
}
BF_variant_group[i,]=variant_group_BF
}
rownames(BF_variant_group)=BF.PP.gene$Gene
colnames(BF_variant_group)=c("(1) PTV or missense variant mpc_>=3", "(2) missense variant mpc_2-3")

BF_variant_group_FDR_gene=as_tibble(BF_variant_group) %>% mutate(Gene=BF.PP.gene$Gene, BF=BF.PP.gene$BF, post.prob=BF.PP.gene$post.prob) %>% filter(Gene %in% FDR_gene$Gene==T)

BF_variant_group_all_gene=as_tibble(BF_variant_group) %>% mutate(Gene=BF.PP.gene$Gene, BF=BF.PP.gene$BF, post.prob=BF.PP.gene$post.prob) %>% arrange(desc(post.prob))

#write.csv(BF_variant_group_all_gene, file="C:\\Users\\hans\\OneDrive - Marquette University\\SCHEMA\\Suppl\\BF_variant_group_all_genes.csv")


BF_variant_group_FDR_gene_plot=data.frame(Gene=c(BF_variant_group_FDR_gene$Gene, BF_variant_group_FDR_gene$Gene), log_BF=c(round(log(BF_variant_group_FDR_gene$`(1) PTV or missense variant mpc_>=3`),2),  round(log(BF_variant_group_FDR_gene$`(2) missense variant mpc_2-3`),2)), group=c(rep("(1) PTV or missense variant mpc_>=3", nrow(BF_variant_group_FDR_gene)), rep("(2) missense variant mpc_2-3", nrow(BF_variant_group_FDR_gene))))

# Create the barplot
ggplot(data=BF_variant_group_FDR_gene_plot[1:20,], aes(x=Gene, y=log_BF, fill=group)) +
  geom_bar(stat="identity")+
  geom_text(aes( label=log_BF), vjust=1.6, angle = 0,
            color="black", size=3.5)+
  scale_fill_brewer(palette="Paired")+
   theme(axis.text.x = element_text(angle = 30, vjust = 0.5, size=7))
  theme_minimal()

```




* overlap with top 10 exome significant genes 


```{r, echo=F, message=F, warning=F}
SCHEMA_top_gene=SCHEMA_gene_results$`Table S5 - Gene Results`$`Gene Symbol`[1:10]
SCHEMA_top_gene2=SCHEMA_gene_results$`Table S5 - Gene Results`[order(SCHEMA_gene_results$`Table S5 - Gene Results`$`P ca/co (comb)`, decreasing = F),]$`Gene Symbol`[1:10]
library("ggvenn")
x=list(
  fdr_gene=FDR_gene$Gene,
  scheme_gene_with_denovo=SCHEMA_top_gene
)
#ggvenn(x)+
#  ggtitle("")+
#  theme(plot.title = element_text(hjust = 0.5, size=20, color = 2)) #center the title 



x=list(
  fdr_gene=FDR_gene$Gene,
  scheme_gene_without_denovo=SCHEMA_top_gene2
)
ggvenn(x)+
  ggtitle("")+
  theme(plot.title = element_text(hjust = 0.5, size=20, color = 2)) #center the title



 fisher.test(matrix(c(10,110, 10,3235), nrow=2))
```


* overlap with top 32 genes at FDR 5\% 


```{r, echo=F, message=F, warning=F}
SCHEMA_top_gene=SCHEMA_gene_results$`Table S5 - Gene Results`$`Gene Symbol`[1:32]
SCHEMA_top_gene2=SCHEMA_gene_results$`Table S5 - Gene Results`[order(SCHEMA_gene_results$`Table S5 - Gene Results`$`P ca/co (comb)`, decreasing = F),]$`Gene Symbol`[1:32]
library("ggvenn")
x=list(
  fdr_gene=FDR_gene$Gene,
  scheme_gene_with_denovo=SCHEMA_top_gene
)
#ggvenn(x)+
#  ggtitle("")+
#  theme(plot.title = element_text(hjust = 0.5, size=20, color = 2)) #center the title 



x=list(
  fdr_gene=FDR_gene$Gene,
  scheme_gene_without_denovo=SCHEMA_top_gene2
)
ggvenn(x)+
  ggtitle("")+
  theme(plot.title = element_text(hjust = 0.5, size=20, color = 2)) #center the title


 fisher.test(matrix(c(22,110, 32,3235), nrow=2))
```




```{r, echo=F, message=F, warning=F}
SCHEMA_top_10_gene_withdenovo=SCHEMA_gene_results$`Table S5 - Gene Results`$`Gene Symbol`[1:10]

SCHEMA_top_10_gene=SCHEMA_gene_results$`Table S5 - Gene Results`[order(SCHEMA_gene_results$`Table S5 - Gene Results`$`P ca/co (comb)`, decreasing = F),]$`Gene Symbol`[1:10] # top 10 genes from case control sample 
SCHEMA_FDR_gene=SCHEMA_gene_results$`Table S5 - Gene Results`[order(SCHEMA_gene_results$`Table S5 - Gene Results`$`P ca/co (comb)`, decreasing = F),]$`Gene Symbol`[1:32]

X=list(
SCHEMA_10_Risk_Gene=SCHEMA_top_gene2,
  MIRAGE_gene=FDR_gene$Gene,
SCHEMA_FDR_gene=SCHEMA_FDR_gene
  )


#library(ggvenn)
#ggvenn(
#  X, 
#  fill_color = c("#0073C2FF", "#EFC000FF",  "#CD534CFF"),
#  stroke_size = 0.5, set_name_size = 4
#  )

#png("C:\\Users\\hans\\OneDrive - Marquette University\\SCHEMA\\Fig_venn_diagram.png")

library("VennDiagram") 
# move to new plotting page 
grid.newpage() 
  
# create Venn diagram with three sets 
draw.triple.venn(area1=10,area2=110, area3 =32,  
                 n12=10, n23=22, n13=10, n123=10,  
                 category=c("SCHEMA 10 risk genes","MIRAGE FDR genes","SCHEMA FDR genes"), 
                 col="red",fill=c("Green","Orange","Blue"), lty="solid", 
                 cat.cex = 1.3, cat.pos=0, cat.col=c("Green","Orange","Blue"), cex=1.5) # cat.cex  the size of the category names; cex: A vector (length 7) giving the size of the areas' labels; cat.pos: position of category names https://www.rdocumentation.org/packages/VennDiagram/versions/1.7.3/topics/draw.triple.venn
#dev.off()
```



```{r, echo=F, message=F, warning=F}
x=list(
  MIRAGE_FDR_gene=FDR_gene$Gene,
  SCHEMA_FDR_gene=SCHEMA_gene_results$`Table S5 - Gene Results`[order(SCHEMA_gene_results$`Table S5 - Gene Results`$`P ca/co (comb)`, decreasing = F),]$`Gene Symbol`[1:33],
  SCZ_gene=SCHEMA_top_10_gene
)
library(VennDiagram)

display_venn <- function(x, ...){
  library(VennDiagram)
  grid.newpage()
  venn_object <- venn.diagram(x, filename = NULL, ...)
  grid.draw(venn_object)
}

display_venn(
        x,
    category.names = c("MIRAGE FDR genes" , "SCHEMA FDR genes", "SCHEMA 10 risk genes"),
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = c("#009E73", "#999999", "#CD534CFF"),
        # Numbers
        cex = 1,
        fontface = "italic",
        # Set names
        cat.cex = 0.8,
        cat.fontface = "bold",
        cat.default.pos = "outer",
       # cat.dist = c(0.055, 0.055, 0.1, 0.1, 0.3 )
         cat.dist = c(0.055, 0.055, 0.1)
)
```





## shared genes with other neurodevelopmental disorders 

```{r, echo=F, message=F, warning=F}
neuro_disorers=SCHEMA_gene_results$`Table S11 - Genes from genome-w`
```

* shared genes with autism 


```{r, echo=F, message=F, warning=F}
Fu_gene_list=as_tibble(read.csv("C:/han/Dataset/20231221_PMID_35982160/Fu 2022 Suppl-41588_2022_1104_MOESM3_ESM.csv"))
ASC_gene=Fu_gene_list %>% filter(ASD185=="TRUE") %>% dplyr::select(gene_gencodeV33) %>% pull



#intersect(FDR_gene$Gene,as_tibble(neuro_disorers) %>% filter(Category=="ASD genes") %>% dplyr::select(`Gene name`) %>% pull())

intersect(FDR_gene$Gene,ASC_gene)


X=list(
  MIRAGE_SCHEMA_gene=FDR_gene$Gene,
Autism_gene=as_tibble(neuro_disorers) %>% filter(Category=="ASD genes") %>% dplyr::select(`Gene name`) %>% pull()
  )


#library(ggvenn)
#ggvenn(
#  X, 
#  fill_color = c("#0073C2FF", "#EFC000FF",  "#CD534CFF"),
#  stroke_size = 0.5, set_name_size = 4
#  )

```


* shared genes with DD/ID genes


```{r, echo=F, message=F, warning=F}
intersect(FDR_gene$Gene, as_tibble(neuro_disorers) %>% filter(Category=="DD/ID genes") %>% dplyr::select(`Gene name`) %>% pull())

X=list(
  MIRAGE_SCHEMA_gene=FDR_gene$Gene,
DD_ID_gene=as_tibble(neuro_disorers) %>% filter(Category=="DD/ID genes") %>% dplyr::select(`Gene name`) %>% pull()
  )


#library(ggvenn)
#ggvenn(
#  X, 
#  fill_color = c("#0073C2FF", "#EFC000FF",  "#CD534CFF"),
#  stroke_size = 0.5, set_name_size = 4
#  )

```


* shared genes with Schizophrenia GWAS genes


```{r, echo=F, message=F, warning=F}
intersect(FDR_gene$Gen, as_tibble(neuro_disorers) %>% filter(Category=="Schizophrenia GWAS genes") %>% dplyr::select(`Gene name`) %>% pull())

X=list(
  MIRAGE_SCHEMA_gene=FDR_gene$Gene,
SCZ_GWAS_gene=as_tibble(neuro_disorers) %>% filter(Category=="Schizophrenia GWAS genes") %>% dplyr::select(`Gene name`) %>% pull()
  )


#library(ggvenn)
#ggvenn(
#  X, 
#  fill_color = c("#0073C2FF", "#EFC000FF",  "#CD534CFF"),
#  stroke_size = 0.5, set_name_size = 4
#  )

```


```{r, echo=F, message=F, warning=F}

Fu_gene_list=as_tibble(read.csv("C:/han/Dataset/20231221_PMID_35982160/Fu 2022 Suppl-41588_2022_1104_MOESM3_ESM.csv"))
ASC_gene=Fu_gene_list %>% filter(ASD185=="TRUE") %>% dplyr::select(gene_gencodeV33) %>% pull

x=list(
  MIRAGE_SCHEMA_gene=FDR_gene$Gene,
  #Autism_gene=as_tibble(neuro_disorers) %>% filter(Category=="ASD genes") %>% dplyr::select(`Gene name`) %>% pull(),
  Autism_gene=ASC_gene,
  DD_ID_gene=as_tibble(neuro_disorers) %>% filter(Category=="DD/ID genes") %>% dplyr::select(`Gene name`) %>% pull(),
  SCZ_GWAS_gene=as_tibble(neuro_disorers) %>% filter(Category=="Schizophrenia GWAS genes") %>% dplyr::select(`Gene name`) %>% pull()
  #SCZ_gene=SCHEMA_top_10_gene
)
library(VennDiagram)

display_venn <- function(x, ...){
  library(VennDiagram)
  grid.newpage()
  venn_object <- venn.diagram(x, filename = NULL, ...)
  grid.draw(venn_object)
}

#png("C:\\Users\\hans\\OneDrive - Marquette University\\SCHEMA\\Fig_overlapping_genes.png")

display_venn(
        x,
    #    category.names = c("MIRAGE FDR genes" , "Autism genes" , "DD/ID genes", "SCZ GWAS genes"),
    category.names = c("MIRAGE FDR genes" , "Autism genes" , "DD/ID genes", "SCZ GWAS genes"),
        # Circles
        lwd = 2,
        lty = 'blank',
     #   fill = c("#999999", "#E69F00", "#56B4E9", "#009E73"),
    fill = c("red", "blue", "green", "yellow"),
        # Numbers
        cex = 1.2,
        fontface = "italic",
        # Set names
        cat.cex = 1.2,
        cat.fontface = "bold",
        cat.default.pos = "outer",
       # cat.dist = c(0.055, 0.055, 0.1, 0.1, 0.3 )
         cat.dist = c(0.055, 0.055, 0.1, 0.1 )
)

#dev.off()
```
```{r, echo=F, message=F, warning=F}
# enrichment analysis 
all_gene=BF.PP.gene$Gene
Autism_gene=ASC_gene
DD_ID_gene=as_tibble(neuro_disorers) %>% filter(Category=="DD/ID genes") %>% dplyr::select(`Gene name`) %>% pull()
SCZ_GWAS_gene=as_tibble(neuro_disorers) %>% filter(Category=="Schizophrenia GWAS genes") %>% dplyr::select(`Gene name`) %>% pull()
SCHEMA_FDR_gene=SCHEMA_gene_results$`Table S5 - Gene Results`[order(SCHEMA_gene_results$`Table S5 - Gene Results`$`P ca/co (comb)`, decreasing = F),]$`Gene Symbol`[1:33]

OR=numeric(); pvalue=numeric()
mirage_percentage=numeric(); all_gene_percentage=numeric(); schema_fdr_percentage=numeric()
####### overlapping with DD/ID_gene 
overlap1=length(intersect(all_gene, DD_ID_gene))
overlap2=length(intersect(FDR_gene$Gene, DD_ID_gene))
overlap3=length(intersect(SCHEMA_FDR_gene, DD_ID_gene))
fisher_test=fisher.test(matrix(c(overlap2, overlap1, length(FDR_gene$Gene), length(all_gene)), nrow=2))
OR[1]=fisher_test$estimate; pvalue[1]=fisher_test$p.value
mirage_percentage[1]=overlap2/length(FDR_gene$Gene)
all_gene_percentage[1]=overlap1/length(all_gene)
schema_fdr_percentage[1]=overlap3/length(SCHEMA_FDR_gene)

####### overlapping with SCZ GWAS gene 
overlap1=length(intersect(all_gene, SCZ_GWAS_gene))
overlap2=length(intersect(FDR_gene$Gene, SCZ_GWAS_gene))
overlap3=length(intersect(SCHEMA_FDR_gene, SCZ_GWAS_gene))
fisher_test=fisher.test(matrix(c(overlap2, overlap1, length(FDR_gene$Gene), length(all_gene)), nrow=2))
OR[2]=fisher_test$estimate; pvalue[2]=fisher_test$p.value
mirage_percentage[2]=overlap2/length(FDR_gene$Gene)
all_gene_percentage[2]=overlap1/length(all_gene)
schema_fdr_percentage[2]=overlap3/length(SCHEMA_FDR_gene)
####### overlapping with Autism gene 
overlap1=length(intersect(all_gene, Autism_gene))
overlap2=length(intersect(FDR_gene$Gene, Autism_gene))
overlap3=length(intersect(SCHEMA_FDR_gene, Autism_gene))
fisher_test=fisher.test(matrix(c(overlap2, overlap1, length(FDR_gene$Gene), length(all_gene)), nrow=2))
OR[3]=fisher_test$estimate; pvalue[3]=fisher_test$p.value
mirage_percentage[3]=overlap2/length(FDR_gene$Gene)
all_gene_percentage[3]=overlap1/length(all_gene)
schema_fdr_percentage[3]=overlap3/length(SCHEMA_FDR_gene)

OR_pvalue=data.frame(OR=OR, pvalue=round(pvalue,4), gene_set=c("DD/ID genes", "SCZ GWAS genes", "Autism genes"))

ggplot(data=OR_pvalue, aes(x=gene_set, y=OR, fill=gene_set)) + 
     geom_bar(position = 'dodge', stat='identity') +
     xlab("")+ylab("odds ratio")+
  theme(legend.position="none")+
     geom_text(aes(label= pvalue), position=position_dodge(width=0.9), vjust=-0.25)

percentage=data.frame(prop=c(all_gene_percentage, mirage_percentage, schema_fdr_percentage), gene_set=c("DD/ID genes", "SCZ GWAS genes", "Autism genes", "DD/ID genes", "SCZ GWAS genes", "Autism genes", "DD/ID genes", "SCZ GWAS genes", "Autism genes"), group=rep(c("all genes", "MIRAGE FDR genes", "SCHEMA FDR genes"), each=3))
topmirage_enrichment=
  ggplot(percentage, aes(gene_set, prop)) + 
   geom_bar(aes(fill = group), stat = "identity", position = "dodge")+
   theme_classic()+
   xlab("")+ylab("Proportion of genes in the gene set")+
  theme(axis.text.y = element_text(size = 10))+ # this is for numbers on y axis 
   ylim(c(0, 0.35))+
   theme(plot.title = element_text(hjust = 0.5, size=15, face="bold"))+
   ggtitle("Enrichment of MIRAGE FDR genes with other gene sets")+
    theme(legend.title = element_blank())+
   geom_signif(annotations ="0.007", y_position = 0.125  ,xmin=0.7, xmax=1.3, size=1, vjust=-0.5, textsize=3.0)+ # textsize is for text over the bar; size is for width of lines 
     geom_signif(annotations ="0.003", y_position = 0.18  ,xmin=1.7, xmax=2.3, size=1, vjust=-0.5, textsize=3.0)+
   geom_signif(annotations ="0.083", y_position = 0.035  ,xmin=2.7, xmax=3.3, size=1, vjust=-0.5, textsize=3.0)+
   theme(legend.position="bottom")+
   theme(axis.text.x = element_text(angle = 0, vjust = 0.5, size=13))+
 theme(text = element_text(size=13))+ # size is for labels on x axis 
  theme(axis.title = element_text(size = 13))+ # Font Size of Axis Titles
 theme(legend.spacing.x = unit(1.0, 'cm')) # add space between symbols in legend 

#png("C:/Users/hans/OneDrive - Marquette University/SCHEMA/Fig-enrichment_of_top_genes.png")
topmirage_enrichment
#dev.off()
  

percentage=data.frame(prop=c(all_gene_percentage, mirage_percentage), gene_set=c("DD/ID genes", "SCZ GWAS genes", "Autism genes", "DD/ID genes", "SCZ GWAS genes", "Autism genes"), group=rep(c("all genes", "MIRAGE FDR genes"), each=3))
topmirage_enrichment=
  ggplot(percentage, aes(gene_set, prop)) + 
   geom_bar(aes(fill = group), stat = "identity", position = "dodge")+
   theme_classic()+
   xlab("")+ylab("Proportion of genes in the gene set")+
  theme(axis.text.y = element_text(size = 10))+ # this is for numbers on y axis 
   ylim(c(0, 0.2))+
   theme(plot.title = element_text(hjust = 0.5, size=15, face="bold"))+
   ggtitle("Enrichment of MIRAGE FDR genes with other gene sets")+
    theme(legend.title = element_blank())+
   geom_signif(annotations ="0.007", y_position = 0.125  ,xmin=0.7, xmax=1.3, size=1, vjust=-0.5, textsize=3.0)+ # textsize is for text over the bar; size is for width of lines 
     geom_signif(annotations ="0.003", y_position = 0.18  ,xmin=1.7, xmax=2.3, size=1, vjust=-0.5, textsize=3.0)+
   geom_signif(annotations ="0.083", y_position = 0.035  ,xmin=2.7, xmax=3.3, size=1, vjust=-0.5, textsize=3.0)+
   theme(legend.position="bottom")+
   theme(axis.text.x = element_text(angle = 0, vjust = 0.5, size=13))+
 theme(text = element_text(size=13))+ # size is for labels on x axis 
  theme(axis.title = element_text(size = 13))+ # Font Size of Axis Titles
 theme(legend.spacing.x = unit(1.0, 'cm')) # add space between symbols in legend
topmirage_enrichment
```


