---
title: "SCHEMA analysis"
output: html_document
date: "2023-09-28"
---


```{r, echo=F, warning=F, message=F}
rm(list=ls())
set.seed(123)
library(knitr)
library(RSQLite)
library(dplyr)
library(knitr)
knitr::opts_chunk$set(cache = TRUE, warning = FALSE, 
                      message = FALSE, cache.lazy = FALSE) # avoid error of long vectors https://stackoverflow.com/questions/39417003/long-vectors-not-supported-yet-error-in-rmd-but-not-in-r-script 
library(kableExtra)
library(RColorBrewer)
library(gplots)
library(tidyverse)
library(gridExtra)
library(ggpubr)
library(DT)
#library("devtools")
#install.packages("Rtools")
#install_github('xinhe-lab/mirage')
library(mirage)

```

[SCHEMA](https://schema.broadinstitute.org/)

```{r, echo=F, message=F, warning=F,cache=T}
SCHEMA_data=as_tibble(read.table("C:\\han\\Dataset\\SCHEMA\\SCHEMA_variant_results.tsv.subset", header=T))
```

## Replicate Extended Data Fig 2  

```{r, echo=F, message=F, warning=F}
MAC_threshold=5 # this cutoff is used in SCHEMA paper in Section Analysis approach  
SCHEMA_syn=SCHEMA_data %>% filter(consequence=="synonymous_variant" & ac_case<=MAC_threshold & ac_ctrl<=MAC_threshold) 

##############
#SCHEMA_syn$ac_case=as.numeric(SCHEMA_syn$ac_case)
#SCHEMA_syn$ac_ctrl=as.numeric(SCHEMA_syn$ac_ctrl)
# what are ac_case, ac_ctrl, an_case, an_ctrl? 
# ac_case: allle count in case; ac_ctrl: allele count in control; an_case: allele number in case; an_ctrl: allele number in control #  see details here: https://schema.broadinstitute.org/gene/ENSG00000055130; move cursor to the variables then their meanings will pop up  


############## check allele frequency 
#max(SCHEMA_syn$ac_case, na.rm=T)
#max(SCHEMA_syn$ac_ctrl, na.rm=T)
#SCHEMA_syn %>% filter(ac_case=="48587") from here, SCHEMA data doesn't use AF filter. 

############### overall Burden of syn mutations 
SCHEMA_syn$ac_case=as.numeric(SCHEMA_syn$ac_case) # convert into numeric value 
SCHEMA_syn$ac_ctrl=as.numeric(SCHEMA_syn$ac_ctrl)
SCHEMA_syn=SCHEMA_syn[complete.cases(SCHEMA_syn$ac_case),] # remove rows when either ac_case or ac_ctrl has NA 
SCHEMA_syn=SCHEMA_syn[complete.cases(SCHEMA_syn$ac_ctrl),]

num_variant=nrow(SCHEMA_syn)
ac_case_total=sum(SCHEMA_syn$ac_case)
ac_ctrl_total=sum(SCHEMA_syn$ac_ctrl)
#num_case=24248; num_ctrl=46885 # numbers from from Fig 1a https://www.nature.com/articles/s41586-022-04556-w#Sec15
#num_case=24248; num_ctrl=50437 # numbers from Table S2
#there are 50437 internal control and 46885 gnomAD controls   
num_case=22444; num_ctrl=39837  # mumbers from Fig 1c, from the same population, see Section-Statistical approach in suppl 

############################ 
OR=numeric(); OR_lower=numeric(); OR_upper=numeric()
#we considered frameshift, stop gained, splice acceptor
#and donor variants as putative protein-truncating (or loss-of-function) variants, so LoF and PTV have the same definition. 
LoF=SCHEMA_data %>% filter(consequence=="frameshift_variant" | consequence=="stop_gained" |consequence=="splice_acceptor_variant" | consequence=="splice_donor_variant")
ac_case_total=sum(as.numeric(LoF$ac_case), na.rm=T); ac_ctrl_total=sum(as.numeric(LoF$ac_ctrl), na.rm=T)
burden_test=fisher.test(matrix(c(ac_case_total,num_case, ac_ctrl_total, num_ctrl), nrow=2))
OR[1]=burden_test$estimate; OR_lower[1]=burden_test$conf.int[1]; OR_upper[1]=burden_test$conf.int[2]


```

## gene CUL1

```{r, echo=F, message=F, warning=F}
CUL1=SCHEMA_data %>% filter(gene_id=="ENSG00000055130" & consequence!="synonymous_variant") # https://schema.broadinstitute.org/gene/ENSG00000055130

#CUL1 %>% filter(locus=="7:148427164") # zoom in this single variant here https://schema.broadinstitute.org/gene/ENSG00000055130
#to know:  (genome): AFR(genomes); (exomes): EUR(exomes); the third one is meta; 
```

